test_util/marathon.py

```diff
@@ -14,91 +17,141 @@
+        if container_port is None:
+            # provide a dummy value for the bridged container port if user is indifferent
+            container_port = 8080
+    else:
+        assert container_port is None or container_port == host_port, 'Cannot declare a different host and '\
+            'container port outside of BRIDGE network'
+        container_port = host_port
+    if network == Network.USER:
+        assert host_port != 0, 'Cannot auto-assign a port on USER network!'
+
     test_uuid = uuid.uuid4().hex
     app = copy.deepcopy({
         'id': TEST_APP_NAME_FMT.format(test_uuid),
         'cpus': 0.1,
         'mem': 32,
         'instances': 1,
         'cmd': '/opt/mesosphere/bin/dcos-shell python '
-               '/opt/mesosphere/active/dcos-integration-test/util/python_test_server.py ',
+               '/opt/mesosphere/active/dcos-integration-test/util/python_test_server.py {}'.format(
+                   # If network is host and host port is zero, then the port is auto-assigned
+                   # and the commandline should reference the port with the marathon built-in
+                   '$PORT0' if host_port == 0 and network == Network.HOST else container_port),
         'env': {
             'DCOS_TEST_UUID': test_uuid,
             # required for python_test_server.py to run as nobody
             'HOME': '/'
         },
         'healthChecks': [
             {
-                'protocol': 'MESOS_HTTP',
+                'protocol': healthcheck_protocol.value,
                 'path': '/ping',
-                'portIndex': 0,
                 'gracePeriodSeconds': 5,
                 'intervalSeconds': 10,
                 'timeoutSeconds': 10,
                 'maxConsecutiveFailures': 3
             }
         ],
     })
-    if not custom_port:
-        app['cmd'] += '$PORT0'
-        app['portDefinitions'] = [{
-            "protocol": "tcp",
-            "port": 0,
-            "name": "test"
-        }]
-    return app, test_uuid
-
-
-def get_test_app_in_docker(ip_per_container=False):
-    app, test_uuid = get_test_app(custom_port=True)
-    assert 'portDefinitions' not in app
-    app['cmd'] += '9080'  # Fixed port for inside bridge networking or IP per container
-    app['container'] = {
-        'type': 'DOCKER',
-        'docker': {
+    if host_port == 0:
+        # port is being assigned by marathon so refer to this port by index
+        app['healthChecks'][0]['portIndex'] = 0
+    elif network == Network.BRIDGE:
+        app['healthChecks'][0]['port'] = container_port if \
+            healthcheck_protocol == Healthcheck.MESOS_HTTP else host_port
+    else:
+        # HOST or USER network with non-zero host port
+        app['healthChecks'][0]['port'] = host_port
+    if container_type != Container.NONE:
+        app['container'] = {
+            'type': container_type.value,
             # TODO(cmaloney): Switch to alpine with glibc
```

Comment:
[request] - Please kill this `TODO` in this one.  My experiments (https://github.com/dcos/dcos/pull/1449) show that using a) docker official alpine image b) alpine+glib image seem to cause failures than the default of `debian:jessie`. I wanted to remove this `TODO` in a separate commit, but I am fine with this being removed as part of this refactor/change.